twenty-first/src/math/polynomial.rs

```diff
@@ -1360,6 +1416,403 @@ where
         let num_coefficients_to_retain = n.min(self.coefficients.len());
         Self::new(self.coefficients[..num_coefficients_to_retain].into())
     }
+
+    /// Preprocessing data for
+    /// [fast modular coset interpolation](Self::fast_modular_coset_interpolate).
+    /// Marked `pub` for benchmarking. Not considered part of the public API.
+    #[doc(hidden)]
+    pub fn fast_modular_coset_interpolate_preprocess(
+        n: usize,
+        offset: BFieldElement,
+        modulus: &Polynomial<FF>,
+    ) -> ModularInterpolationPreprocessingData<FF> {
+        let omega = BFieldElement::primitive_root_of_unity(n as u64).unwrap();
+        // a list of polynomials whose ith element is X^(2^i) mod m(X)
+        let modular_squares = (0..n.ilog2())
+            .scan(
+                Polynomial::<FF>::new(vec![FF::from(0), FF::from(1)]),
+                |acc, _| {
+                    let yld = acc.clone();
+                    *acc = acc.multiply(acc).reduce(modulus);
+                    Some(yld)
+                },
+            )
+            .collect_vec();
+        let even_zerofier_lcs = (0..n.ilog2())
+            .map(|i| offset.inverse().mod_pow(1u64 << i))
+            .collect_vec();
+        let even_zerofiers = even_zerofier_lcs
+            .into_iter()
```

Comment:
There's at least two superfluous allocations happening in these and the following lines. For example, `even_zerofier_lcs` is `.collect()`ed into a vector, which is then directly consumed by `.into_iter()`.

I don't know whether the compiler takes care of this.

```suggestion
        let even_zerofier_lcs = (0..n.ilog2())
            .map(|i| offset.inverse().mod_pow(1u64 << i));
        let even_zerofiers = even_zerofier_lcs
```

Reply:
306fa9ae